js/modules/k6/experimental/fs/module.go

```diff
@@ -186,8 +186,31 @@ func (f *File) Stat() *goja.Promise {
 //
 // It is possible for a read to successfully return with 0 bytes.
 // This does not indicate EOF.
+//
+//nolint:funlen
 func (f *File) Read(into goja.Value) *goja.Promise {
-	promise, resolve, reject := promises.New(f.vu)
+	// This method performs an asynchronous read operation and modifies the provided Uint8Array in place.
+	// To ensure thread safety and avoid concurrency issues, we take special precautions when creating the promise:
+	//
+	// 1. Instead of using the standard [promises.New] method, we manually create a promise.
+	// 2. We register a callback to be executed by the VU's runtime. This ensures that the modification
+	//    of the JS runtime's `buffer` occurs on the main thread during the promise's resolution.
+	promise, resolveFunc, rejectFunc := f.vu.Runtime().NewPromise()
+	callback := f.vu.RegisterCallback()
+
+	resolve := func(result any) {
+		callback(func() error {
+			resolveFunc(result)
+			return nil
+		})
+	}
+
+	reject := func(reason any) {
+		callback(func() error {
+			rejectFunc(reason)
+			return nil
+		})
+	}
```

Comment:
You can use the helper you create [here](https://github.com/grafana/k6/blob/61db2fea4b3bdaf73e0505aa7b6fd5c07d650117/js/promises/promises.go#L30C7-L30C7)

Reply:
I considered it as of the writing of the code, but I was not sure how it would be possible to do https://github.com/grafana/k6/pull/3420/files#diff-879feac1c36a4a6578f2509b47bf5dab13a5a93d89a05e9dcf4c71613758d19dR254 then.

Namely, the helper doesn't give me access to the underlying goja resolving function. Also, would it be okay to declare a "side" callback manually, side-by-side with the one created under the hood for the `resolve` and `reject` functions returned by `promises.New`? 😃 ❓

Reply:
You are correct, and having now gone through the code I saw other problems. 

But let's merge this now and I will point them out and we will fix them in the PRs that aren't about the races.

Reply:
Sounds good 🙇🏻